test/integration/http_integration.cc

```diff
@@ -936,6 +936,57 @@ void HttpIntegrationTest::testIdleTimeoutWithTwoRequests() {
   test_server_->waitForCounterGe("cluster.cluster_0.upstream_cx_idle_timeout", 1);
 }
 
+void HttpIntegrationTest::testUpstreamDisconnectWithTwoRequests() {
+  initialize();
+  fake_upstreams_[0]->set_allow_unexpected_disconnects(true);
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Request 1.
+  codec_client_->makeRequestWithBody(Http::TestHeaderMapImpl{{":method", "GET"},
+                                                             {":path", "/test/long/url"},
+                                                             {":scheme", "http"},
+                                                             {":authority", "host"}},
+                                     1024, *response_);
+  waitForNextUpstreamRequest();
+
+  // Request 2.
+  IntegrationStreamDecoderPtr response2{new IntegrationStreamDecoder(*dispatcher_)};
+  IntegrationCodecClientPtr codec_client2 = makeHttpConnection(lookupPort("http"));
+  codec_client2->makeRequestWithBody(Http::TestHeaderMapImpl{{":method", "GET"},
+                                                             {":path", "/test/long/url"},
+                                                             {":scheme", "http"},
+                                                             {":authority", "host"}},
+                                     512, *response2);
+
+  // Response 1.
+  upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+  upstream_request_->encodeData(512, true);
+  fake_upstream_connection_->close();
```

Comment:
Sorry, I may not have had sufficient caffeine, but how are we making sure that the test Envoy is receiving the FIN before assigning the next request?  If there were a context switch between encodeData and the close() couldn't Envoy read the whole request and reassign the upstream before the close() occurs?    We may need to unit test this rather than integration test.  Alternately we could do a raw tcp connection, send the response and stray data in one write call.  I think that'd guarantee no race and I'd hope any activity on the delayed-use connection would cause it to be removed.

Reply:
The dispatcher is not running until next wait* call, so in the right next line `waitForEndSream` the dispatcher runs, the connection read 512 bytes and FIN in same event. So the request is not being scheduled before close() occurs.

Also added a unit test in conn_pool_test too.

Reply:
Yeah, I'd like to think this will be fine, we've just had a spate of macos integration flakes due to slightly different network semantics.  We can just keep an eye out for test flakes and remove this if it causes problems now that we have a unit test.  @zuercher just so it's on his radar.